### DISCUSSION TITLE
CSV Synchronisation issues

### DISCUSSION THREAD

**Discussion by ghost** [jspsych/jsPsych#2513] [General]
I am really having trouble saving csv responses on my server and I don't understand why. I have attached a screenshot of my php code as well as a list of qualtrics responses with times (the qualtrics survey redirects to jspsych after the end of the survey). The last modified date for the csv files was when I last looked at the files, and the last id recorded was R_2YQ. I don't understand why this isn't working :( I have tested it on my own computer and it has worked fine.  NB I appended 2 to the file names to see if it creates a new file for the next participant
This is the php code and saving function which should append data to the csv files. 

<img width="886" alt="phpcode" src="https://user-images.githubusercontent.com/71668806/154007076-11dd8c11-1265-4622-9380-01bfbee53910.png">

<img width="636" alt="save_data" src="https://user-images.githubusercontent.com/71668806/154007017-79e07bd5-a0d5-4238-b793-0b18b86740a9.png">

<img width="886" alt="phpcode" src="https://user-images.githubusercontent.com/71668806/154006023-cf875663-3653-4f1f-935c-f5
<img width="937" alt="qualtricsmissingresponses" src="https://user-images.githubusercontent.com/71668806/154006475-8026cd3c-ad99-4871-9ff7-98cacd7a3cdb.png">
376b8d8a5d.png">
<img width="884" alt="lastmodified" src="https://user-images.githubusercontent.com/71668806/154006273-7f9761d2-2c15-44a5-a653-c415d58fa503.png">
Thank you

**Comment by becky-gilbert** [comment]
Hi @Vizitok, I'm not a PHP person so I can't comment on that side of things, but one possibility is that your participants are closing the browser tab/window before the data has been sent back to the server. To get around that problem, you can try sending the data earlier in the experiment. For instance, you could move the code that's currently in your `jsPsych.init` `on_finish` function into the `on_start` or `on_finish` function for the last trial in your experiment. You could also put the code inside of a `call-function` trial that is placed before the last trial in your timeline.

Ideally your experiment wouldn't actually show the participant a "Thanks, you're done!" page until the data has been successfully sent back to the server. To do this, you can use the `call-function` plugin's `async` option, as well as the `onload` callback function for the `XMLHttpRequest`.  This is explained in this section of the jsPsych docs: https://www.jspsych.org/7.1/overview/data/#storing-data-permanently-in-a-mysql-database

I hope this helps. Please let us know if this ends up being the problem, or if it's something else.
